### DISCUSSION TITLE
Sound - Need a few pointers and guides.

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#1032] [Q&A]
@maluoi Q1: Do sound plays span frames? If I wanted to create a metronome in HoloLens 2, where the "tick" sound is a fixed length in milliseconds and spaced evenly across frames, how would something like this be accomplished using a generated sound? Or on the other extreme, I wanted to play generated ticks .5 millisecond duration and I wanted to have a variable number of ticks in per frame in successive frames, how do you do something like this.

I have been experimenting with the sound generation, the Time.Step method, and the Windows Tick time but so far haven't accomplished my objectives.

So far I have been using the SK Demo for Sounds and the SK documentation for sound generation.

**Comment by maluoi** [comment]
I'd probably recommend generating a stream of clicks, where maybe you buffer in a collection of clicks, and then just extend that buffer as time goes on and you start to run out!

I'd recommend taking a look at the little tone-generator radio demo I've got. This basically just generates a tone, and allows you to adjust it with a dial on the "radio" model.
https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/Demos/DemoNodes.cs

**Comment by laultman** [comment]
@maluoi So what is the relationship to Time and frames and the Sound.Play method. As I understand it, in HoloLens 2 the frame rate is non-deterministic and Time.Step methods are only evaluated at a frame Step leading edge. From my testing once the frame starts there is no "intra" frame time measurement capability, is this true? If these are true, then its seems that I must use the Windows Time outside the frame step to measure the passage of time. That would seem to indicate that somehow during a frame I would have to sample the Time ticks.
That is why I was asking about the sound spanning a frame. If it turns out that I need to start a sound in one frame with less than a millisecond "left" before the frame ends (outside my control), does the sound continue to play "between" frames? FYI this has other uses in the industrial controls space that is highly dependent on time intervals. My experiments with sound are good stand-ins for industrial controls

**Comment by laultman** [comment]
@maluoi The "radio" demo looks promising! I wish there were some mode that could be switched on or off to put the render engine into a deterministic frame rate. The entire world of industrial controls uses defined deterministic operations with high accuracy clocks.

Edit: Your ideas about the below are valuable to me. I am however looking into using your existing Time classes to try and make something more deterministic. I see the Time.Frame and the Time.Step, Time.Total as possible ways to implement a time-base.


An alternative would be some kind of time-marker method that returns the current app render frame as a long int, and the current time tick while in a frame. At least we could then do some time management within the context of a single frame. HoloLens 2 does have access to the 100-nanosecond clock but it has no way to know where it is in the relation to the frame sequence.

I think about an app as an old-time movie where each frame is played to the user, one after the other. The frames in a movie are projected at a specific frame rate regardless of the content in the frame. The audio track in the movie is alongside the picture frames and is a continuous stream of audio. The synchronization (audio/video frame) is then maintained by the physical media. It seems that a variable frame rate cannot be synchronized internally and them must have some external mechanism for this. It seems that a simple frame counter would allow you to achieve the external sync.
I was able to do this in a limited way with SK and mp4 video by converting the mp4 to png "frames" and setting up a clock to "play" the frames at a defined frame rate. Because each png was a sequenced frame, I could move forward backward or do random access.

In the case of a stream of frames, the context of the action would have to be an app issue not an SK issue. SK could just return a frame number and a time stamp at the top of the frame. Sounds simple, but I know it is a deep issue. However, SK does "know" because the first registered IStepper is the top of the stack.

Thanks, Nick. Just throwing out some ideas.

**Reply by maluoi** [reply]
StereoKit itself can't guarantee a fixed frame rate, as it depends on the performance of a wide variety of different systems, the OpenXR compositor, GPU work, the apps' work, etc. StereoKit simply goes as fast as it can within the refresh rate of the device.

If you need a more specific time step, you can do a few things:

1. Drive your own time step with a fixed value. This is common for physics simulations, as it allows for smaller, more frequent time steps than what the rendering loop provides! Basically you would have a fixed time step value you want to iterate at, and a timer for how far you've advanced. Then each StereoKit frame, you loop your fixed time step until it has caught up with StereoKit's time.
2. Use a separate thread. This is how audio usually works, since the "refresh rates" for audio are much faster. In this case, you can bypass SK's time related APIs entirely, and just work with C#'s APIs. Just be sure to sleep your thread if you have spare time, otherwise you can run your CPUs to 100% quite easily.

If you are doing visual work though, it's best to work within the constraints of the display loop's timing, trying to bypass that will only result in timing related visual artifacts.

If you're trying to do a flipbook timer, you might do something like this. This approach is particularly great if your flipbook frames may have a variable duration, such as with a .gif.

```csharp
int flipFrame;
float flipTimer;
float flipFrameDuration = 1/30.0f;

void StartFlipbook() {
    flipTimer = Time.Totalf;
    flipFrame = 0;
}

void Step() {
    while ((Time.Totalf - flipTimer) > flipFrameDuration) {
        flipTimer += flipFrameDuration;
        flipFrame += 1;
    }
    
    DrawFrame(flipFrame);
}
```

If you have a fixed frame duration like in a video, then you can simplify. This is how I do basically all of my own timers in StereoKit.

```csharp
float flipTimer;
float flipFrameDuration = 1/30.0f;

void StartFlipbook() {
    flipTimer = Time.Totalf;
}

void Step() {
    int flipFrame = (Time.Totalf - flipTimer) / flipFrameDuration;

    DrawFrame(flipFrame);
}
```

**Comment by laultman** [comment]
@maluoi Continuing to research the deterministic time issues. Ultimately my goal is to create field controls for systems that currently are using centralized command and control systems.

I have been running through the C++ to see what is already there. On the surface it looks like you have 95% of the pieces there to reach a deterministic (or at least within tolerance) time functionality. The sokol_time.h seems to have the necessary functions. The stereokit.cpp has the struct sk_state_t that looks like it could hold a value for the system clock ticks. I can see in the sk_step_begin() method that you are calling the appropriate time method already: local.app_system->profile_frame_start = stm_now();

The question is, since this method is already called at the start of each frame, could you not add a variable to the existing struct to hold this value and to surface it the same way you currently do the Time.Frame value?

This would give me a frame reference and a time value that is tied to the system clock.
